synapse/lib/stormlib/vault.py

```diff
@@ -0,0 +1,738 @@
+        return vault.get(self._vault_field_name)
+
+class VaultSecrets(VaultConfigs):
+    _vault_field_name = 'secrets'
+    _vault_perm = s_cell.PERM_EDIT
+
+    async def setitem(self, name, valu):
+        vault = self.runt.snap.core.reqVault(self.valu)
+
+        name = await s_stormtypes.tostr(name)
+
+        if valu is s_stormtypes.undef:
+            valu = s_common.novalu
+        else:
+            valu = await s_stormtypes.toprim(valu)
+
+        return await self.runt.snap.core.setVaultSecrets(self.valu, name, valu)
+
+@s_stormtypes.registry.registerType
+class Vault(s_stormtypes.Prim):
+    '''
+    Implements the Storm API for a Vault.
+
+    Callers (instantiation) of this class must have already checked that the user has at least
+    PERM_READ to the vault.
+    '''
+    _storm_locals = (
+        {'name': 'iden', 'desc': 'The Vault iden.', 'type': 'str', },
+        {'name': 'type', 'desc': 'The Vault type.', 'type': 'str', },
+        {'name': 'scope', 'desc': 'The Vault scope.', 'type': 'str', },
+        {'name': 'owner', 'desc': 'The Vault owner (user or role iden).', 'type': 'str', },
+        {'name': 'permissions',
+         'desc': 'The Vault permissions.',
+         'type': {
+             'type': ['gtor'],
+             '_gtorfunc': '_gtorPermissions',
+             'returns': {'type': 'dict'}}},
+        {'name': 'configs',
+         'desc': 'The Vault configs data.',
+         'type': {
+             'type': ['gtor'],
+             '_gtorfunc': '_gtorConfigs',
+             'returns': {'type': 'vault:data'}}},
+        {'name': 'secrets',
+         'desc': 'The Vault secrets data.',
+         'type': {
+             'type': ['gtor'],
+             '_gtorfunc': '_gtorSecrets',
+             'returns': {'type': 'vault:data'}}},
+
+        {'name': 'name',
+         'desc': 'The Vault name.',
+         'type': {
+             'type': ['gtor', 'stor'],
+             '_storfunc': '_storName',
+             '_gtorfunc': '_gtorName',
+             'returns': {'type': 'str'}}},
+
+        {'name': 'setPerm', 'desc': 'Set easy permissions on the Vault.',
+         'type': {'type': 'function', '_funcname': '_methSetPerm',
+                  'args': (
+                      {'name': 'iden', 'type': 'str', 'desc': 'The user or role to modify.'},
+                      {'name': 'level', 'type': 'str', 'desc': 'The easyperm level for the iden. $lib.undef to remove an existing permission.'},
+                  ),
+                  'returns': {'type': 'boolean', 'desc': '$lib.true if the permission was set, $lib.false otherwise.', }}},
+
+        {'name': 'delete', 'desc': 'Delete the Vault.',
+         'type': {'type': 'function', '_funcname': '_methDelete',
+                  'args': (),
+                  'returns': {'type': 'boolean', 'desc': '$lib.true if the vault was deleted, $lib.false otherwise.', }}},
+    )
+    _storm_typename = 'vault'
+    _ismutable = False
+
+    def __init__(self, runt, valu, path=None):
```

Comment:
Did you consider the use of fully overwriting the configs/secrets for an existing vault?  AFAICT this would require individually setting/$lib.undef'ing each item in those objects, or deleting and re-creating the vault.

Reply:
We discussed this at scrum the other day and I think there's a separate ticket for it so resolving this conversation.